### PR TITLE
Add Lock and Unlock methods to Client

### PR DESCRIPTION
**Pull request by annelo-msft**
Sample was requested for Lock/Unlock functionality, per https://github.com/Azure/azure-sdk-for-net/issues/7402

We don't currently support these, so adding these methods in addition to the sample.

### CODE REVIEW

**Comment by annelo-msft** [comment]
Remaining to-do: 
- [ ] Add Live/Recorded tests
- [ ] Add Mock test for ETag/If-Match Header

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs [hunk 1]
```diff
@@ -785,5 +785,195 @@ private Request CreateGetRevisionsRequest(SettingSelector selector, string pageL
...
+        /// <summary>
+        /// Locks an existing <see cref="ConfigurationSetting"/> in the configuration store.
+        /// </summary>
+        /// <param name="key">The primary identifier of a configuration setting.</param>
+        /// <param name="label">The value used to group configuration settings.</param>
+        /// <param name="etag">The value of an etag indicates the state of a configuration setting within a configuration store.
+        /// If it is specified, the configuration setting is only locked if etag value matches etag value in the configuration store.
+        /// If no etag value is passed in, then the setting is always locked.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
+        public virtual async Task<Response> LockAsync(string key, string label = default, ETag etag = default, CancellationToken cancellationToken = default)
```
Comment (pakrym) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs]: No overload that takes `ConfigurationSetting`?

File: sdk/appconfiguration/Azure.Data.AppConfiguration/samples/Sample3_LockUnlockSetting.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+using System;
+using NUnit.Framework;
+
+namespace Azure.Data.AppConfiguration.Samples
+{
+    [Category("Live")]
```
Comment (pakrym) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/samples/Sample3_LockUnlockSetting.cs]: I think we have a new attribute for live tests.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs [hunk 3]
```diff
@@ -785,5 +785,195 @@ private Request CreateGetRevisionsRequest(SettingSelector selector, string pageL
...
+                Response response = _pipeline.SendRequest(request, cancellationToken);
+
+                switch (response.Status)
+                {
+                    case 200:
+                        return response;
+                    case 404:
+                        throw response.CreateRequestFailedException("Not found.");
+                    case 412:
+                        throw response.CreateRequestFailedException("The specified precondition failed.");
```
Comment (pakrym) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs]: How does the default error message look in this case? It includes the response phrase so maybe we don't need to include description text.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/samples/Sample3_LockUnlockSetting.cs (hunk 2 above)
Comment (heaths) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/samples/Sample3_LockUnlockSetting.cs]: Only for tests that already take advantage of the recording infrastructure, though. IIRC, appconfig samples didn't use it.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs [hunk 4]
```diff
@@ -785,5 +785,195 @@ private Request CreateGetRevisionsRequest(SettingSelector selector, string pageL
...
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
+        public virtual async Task<Response> UnlockAsync(string key, string label = default, ETag etag = default, CancellationToken cancellationToken = default)
+        {
+            using DiagnosticScope scope = _pipeline.Diagnostics.CreateScope("Azure.Data.AppConfiguration.ConfigurationClient.Unlock");
+            scope.AddAttribute("key", key);
+            scope.Start();
+
+            try
+            {
+                using Request request = CreateUnlockRequest(key, label, etag);
```
Comment (heaths) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs]: Duplicated quite a bit - mainly concerned with changes to the response handling. Perhaps refactor into a helper function?

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs [hunk 5]
```diff
@@ -785,5 +785,195 @@ private Request CreateGetRevisionsRequest(SettingSelector selector, string pageL
...
+            {
+                scope.Failed(e);
+                throw;
+            }
+        }
+
+        private Request CreateLockRequest(string key, string label, ETag etag)
+        {
+            if (string.IsNullOrEmpty(key))
+                throw new ArgumentNullException(nameof(key));
```
Comment (heaths) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs]: I have some helpers coming out for this that will be in Azure.Core, but only if `key` is null should you throw `ArgumentNullException`. If empty, that should be `ArgumentException`. Could give the wrong idea, and this is consistent with the Framework/Core. You could also use Azure.Core's `Guard` class now and I'll be fixing it up later.

**Review by heaths — APPROVED** [review]

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs (hunk 1 above)
Comment (annelo-msft) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs]: We have an open issue about overloads that I need to get resolved.  Once we close on it, I'll update or not, based on the outcome.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs (hunk 3 above)
Comment (annelo-msft) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs]: Per discussion with the feature team, we're going to use the default error response.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs (hunk 4 above)
Comment (annelo-msft) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs]: Good point.  I've taken out special handling over error codes.  The remaining is consistent with how we're implementing response handling in the rest of the file.  Let me know if you still have concerns.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs (hunk 5 above)
Comment (annelo-msft) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs]: @heaths, could you file an issue for me to update this once your tools become available?  We're handling this incorrectly elsewhere in the library and I'd like to fix it in one swoop.

**Comment by annelo-msft** [comment]
Finalizing without incorporating updated board guidance on ETags.

**Pull request merged**